internal/graph/check.go

```diff
@@ -984,6 +991,491 @@ func trySendUsersetsAndDeleteFromMap(ctx context.Context, usersetsMap usersetsMa
+
+		for {
+			t, err := filteredIter.Next(ctx)
+			if err != nil {
+				if errors.Is(err, storage.ErrIteratorDone) || errors.Is(err, context.Canceled) || errors.Is(err, context.DeadlineExceeded) {
+					break
+				}
+
+				// error encountered.  No need to process further
+				span.RecordError(err)
+				concurrency.TrySendThroughChannel(ctx, usersetMessage{
+					userset: "",
+					err:     err,
+				}, usersetMessageChan)
+				return
+			}
+			usersetName, relation := tuple.SplitObjectRelation(t.GetUser())
+			if relation == "" {
+				// This should never happen because the filter should only allow userset with relation.
+				err = fmt.Errorf("unexpected userset %s with no relation", t.GetUser())
+				span.RecordError(err)
+				concurrency.TrySendThroughChannel(ctx, usersetMessage{
+					userset: "",
+					err:     err,
+				}, usersetMessageChan)
+				return
+			}
+			concurrency.TrySendThroughChannel(ctx, usersetMessage{
+				userset: usersetName,
+				err:     nil,
+			}, usersetMessageChan)
+		}
+	}()
+	return usersetMessageChan
+}
+
+func processUsersetMessage(message usersetMessage,
+	primarySortedSet storage.SortedSet,
+	secondarySortedSet storage.SortedSet) (bool, error) {
+	if message.err != nil {
+		return false, message.err
+	}
+	primarySortedSet.Add(message.userset)
+	if secondarySortedSet.Exists(message.userset) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func matchUsersetFromUserAndUsersetFromObject(ctx context.Context,
+	req *ResolveCheckRequest,
+	userToUsersetMessageChan,
+	objectToUsersetMessageChan chan usersetMessage) (*ResolveCheckResponse, storage.SortedSet, storage.SortedSet, error) {
+	usersetFromUser := storage.NewSortedSet()
+	usersetFromObject := storage.NewSortedSet()
+
+	userToUsersetDone := false
+	objectToUsersetDone := false
+
+	for !userToUsersetDone || !objectToUsersetDone {
+		select {
+		case <-ctx.Done():
+			return nil, nil, nil, ctx.Err()
+		case userToUsersetMessage, ok := <-userToUsersetMessageChan:
+			if !ok {
+				userToUsersetDone = true
+				if usersetFromUser.Size() == 0 {
+					return &ResolveCheckResponse{
+						Allowed: false,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							// It probably is +1.  However, we have no control on whether
+							// the userset member lookup for the first level has started DS query.
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+			} else {
+				found, err := processUsersetMessage(userToUsersetMessage, usersetFromUser, usersetFromObject)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+				if found {
+					return &ResolveCheckResponse{
+						Allowed: true,
+						ResolutionMetadata: &ResolveCheckResponseMetadata{
+							DatastoreQueryCount: req.GetRequestMetadata().DatastoreQueryCount + 2,
+						},
+					}, nil, nil, nil
+				}
+				//  check to see if there is a direct assignment
+				if req.GetTupleKey().GetObject() == userToUsersetMessage.userset {
```

Comment:
I wonder if we should also modify `checkDirectUserTuple` so that if we are running the `nestedUsersetFastpath` approach we don't need run `checkDirectUserTuple` because we have line 1387.

Reply:
Probably check direct is better as it covers the non-nested case.